Feat/wallet feature

vineyy17 opened the pull request:
Wallet feature

Onyelaudochukwuka commented on src/utils/api_helper.ts:
```diff
@@ -1656,3 +1671,194 @@ export const requestInvoice = async (
+    ]
+
+    // Get all token configurations from chains.ts
+    const tokenConfigs = [
```
Why isn't this being dynamically pulled? All the data defined here already exists on the "supportedChains.acceptableTokens". With this setup, we need to update tokens in two places when the only thing we have changing is the chains.

Onyelaudochukwuka replied:
Dynamically pulling the data should also ensure that the token decimal is always in context

Onyelaudochukwuka commented on src/utils/database.ts:
```diff
@@ -4407,7 +4570,12 @@ const createCryptoTransaction = async (
     total_fee: feeInUSD,
-    metadata: {},
+    metadata: {
+      sender_address: account_address.toLowerCase(),
```
Why are we storing, account_address twice ? it's already being stored in the intiator address yes ?

Onyelaudochukwuka commented on src/utils/database.ts:
```diff
@@ -4417,46 +4585,78 @@ const createCryptoTransaction = async (
+        .update({
+          metadata: {
+            sender_address: account_address.toLowerCase(),
+            receiver_address: meetingType.account_owner_address.toLowerCase(),
```
This is misleading; the receiver address could be a custom address, so doing this invalidates such actions. The transaction_hash should be the only source of truth

Onyelaudochukwuka commented on src/utils/database.ts:
```diff
@@ -4407,7 +4570,12 @@ const createCryptoTransaction = async (
+    metadata: {
+      sender_address: account_address.toLowerCase(),
+      ...(transactionRequest.receiver_address && {
+        receiver_address: transactionRequest.receiver_address.toLowerCase(),
```
The receiver address here shouldn't be client-dependent. What I expect you to do is modify "getTransactionFeeThirdweb" to get the receiver address and use that here.

Onyelaudochukwuka commented on src/utils/walletConfig.ts:
```diff
@@ -0,0 +1,194 @@
+]
+
+// Centralized network configuration using chains.ts
+export const NETWORKS: Network[] = [
```
This should be based on the environment, I.e, supportedPaymentChains on the constant/meeting-types file

Onyelaudochukwuka commented on src/utils/walletConfig.ts:
```diff
@@ -0,0 +1,194 @@
+    name: 'Celo Dollar',
+    symbol: 'cUSD',
+    icon: '/assets/tokens/CUSD.png',
+    price: '1 USD',
```
The price for stablecoins is never constant; use an oracle like Chainlink to always dynamically pull this information. (services/chainlink.service.ts)

Onyelaudochukwuka commented on src/hooks/useCryptoBalance.ts:
```diff
@@ -0,0 +1,60 @@
+    },
+    enabled: !!currentAccount?.address && !!tokenAddress && chainId !== 0,
+    staleTime: 30000,
+    cacheTime: 60000,
```
We don't want to cache balances. They need to be as real-time as possible, and the stale value from caching doesn't help with that.

Onyelaudochukwuka commented on src/hooks/useCryptoBalances.ts:
```diff
@@ -0,0 +1,184 @@
+    getChainId(SupportedChain.ARBITRUM)
+  )
+
+  const usdcArbitrumSepoliaBalance = useCryptoBalance(
```
We want to expand to more chains in the future, so having a manual approach would snowball into a lot of code quickly. What we want is a way to define just the chain and pull the token information based on what we have defined in the supportedChains variables. We also don't want to need to wait for 6 variables to do the same thing. We should instead run all these fetches in parallel by having a functional-based approach

Onyelaudochukwuka commented on src/hooks/useCryptoBalances.ts:
```diff
@@ -0,0 +1,184 @@
+          isLoading: usdtArbitrumBalance.isLoading,
+        },
+      ]
+    } else if (selectedNetwork === 'Arbitrum Sepolia') {
```
Manual filtering like this isn't an efficient way to address this, when we have 10 chains we then have 10 different branches of IFs ?

Onyelaudochukwuka commented on src/types/chains.ts:
```diff
@@ -78,6 +81,73 @@ export const getTokenIcon = (token: AcceptedToken) => {
   }
 }
 
+export const getTokenDecimals = (token: AcceptedToken): number => {
```
Token decimals shouldn't be statically defined; they should be based on the token information pulled. I made mention of this in a comment above

vineyy17 committed:
refactor: Enhance SendFundsModal and NavMenu for improved token handling and dynamic network configuration

- Removed hardcoded token filtering logic in SendFundsModal, replacing it with dynamic token retrieval based on chain configuration.
- Updated NavMenu to support dropdown items for Payments, improving navigation structure.
- Refactored crypto balance hooks to utilize new token service functions for fetching balances and decimals.
- Introduced a new useCryptoConfig hook for fetching dynamic crypto asset configurations.
- Adjusted wallet balance and transaction hooks for better performance and consistency.

vineyy17 committed:
refactor: Update useCryptoBalances and walletConfig for enhanced token information retrieval

- Improved useCryptoBalances to dynamically fetch token name, symbol, and icon using new utility functions.
- Updated walletConfig interface to include optional error handling for better state management.

vineyy17 committed:
Merge develop

vineyy17 committed:
feat: Enhance profile settings and notifications management

- Introduced a new Settings component to consolidate account details, notifications, and billing sections.
- Added AccountPlansAndBilling and WalletAndPayment components for managing subscriptions and payment settings.
- Improved DiscordNotificationConfig and NotificationConfig components for better user experience and layout.
- Implemented TransactionPinModal for secure transaction management.
- Updated NavMenu to reflect new settings structure and navigation flow.

vineyy17 committed:
feat: Implement payment preferences and transaction PIN management

- Added functionality for managing payment preferences, including saving and retrieving user settings.
- Introduced transaction PIN features, allowing users to create, change, and disable their PIN securely.
- Integrated modals for PIN management and reset processes, enhancing user experience and security.
- Updated API routes for handling payment preferences and PIN verification.
- Enhanced the WalletAndPayment and TransactionPinModal components to support new features.

vineyy17 committed:
feat: Add email change and transaction PIN features

- Implemented functionality for changing account email, including secure email verification via magic links.
- Introduced modals for email change confirmation and transaction PIN management, enhancing user experience.
- Updated API routes for sending email change and PIN enable/reset links.
- Enhanced profile components to support new email and PIN features, ensuring better user interaction and security.
- Added email templates for change email and enable/reset PIN notifications.

vineyy17 committed:
feat: Implement JWT-based email change and PIN management

- Added JWT authentication for email change and PIN enable/reset processes, enhancing security with token expiration.
- Updated email templates to reflect a 5-minute expiration for magic links, improving user awareness.
- Refactored API routes to utilize JWT for secure token generation and verification.
- Enhanced frontend components to handle token validation and user feedback for email and PIN actions.
- Introduced utility functions for JWT verification and expiration checks, streamlining token management.

vineyy17 committed:
feat: Revamp SendFundsModal for transaction verification

- Replaced PIN verification with a two-step verification process involving a PIN and a verification code for enhanced security.
- Introduced a new TransactionVerificationModal component to handle the verification flow.
- Updated state management to reflect changes in verification modal visibility and loading states.
- Modified API routes to support sending and verifying the new verification code.
- Enhanced email helper functions to send verification codes to users, improving the overall security of fund transactions.

vineyy17 committed:
feat: Refactor Wallet component and related hooks for improved chain and token management

- Updated Wallet component to utilize chain IDs for network selection, enhancing compatibility with dynamic chain configurations.
- Refactored useCryptoBalances hook to support selectedChain instead of selectedNetwork, streamlining token balance retrieval.
- Enhanced WalletProvider to initialize selectedNetwork based on wallet-supported chains, improving user experience.
- Modified walletConfig to dynamically fetch token information based on supported chains, ensuring accurate token display and management.
- Added walletSupported flag to chain configurations for better handling of supported tokens across different networks.

vineyy17 committed:
refactor: Update SendFundsModal and Wallet components for improved network handling

- Changed selectedNetwork prop in SendFundsModal to require a SupportedChain type, enhancing type safety.
- Simplified state initialization for sendNetwork in SendFundsModal to directly use selectedNetwork.
- Updated Wallet component to pass selectedNetwork directly, improving clarity and reducing unnecessary logic.
- Enhanced WalletProvider to fetch and set user's preferred network based on payment preferences, ensuring a better user experience.

vineyy17 committed:
feat: Integrate verification code functionality in SendFundsModal

- Added verifyVerificationCode function to streamline verification code handling.
- Updated SendFundsModal to utilize the new verification function, enhancing code clarity and maintainability.
- Improved type safety by specifying types for notification subscriptions in SendFundsModal.
- Removed redundant API call code, simplifying the verification process.

vineyy17 committed:
feat: Enhance responsiveness and styling in profile modals and wallet component

- Updated ReceiveFundsModal and SendFundsModal to improve responsiveness with dynamic sizing and spacing based on screen size.
- Refactored Wallet component to enhance layout and styling, ensuring better user experience across devices.
- Adjusted font sizes and padding for various elements to maintain consistency and improve readability.
- Implemented responsive design principles in MagicLinkModal and WalletActionButton for better adaptability.

vineyy17 committed:
feat: Implement email notifications for crypto transactions and session bookings

- Added functionality to send crypto debit notification emails to users upon successful transactions.
- Introduced session booking income notification emails to inform hosts of received payments.
- Enhanced email templates for better user experience and clarity in transaction details.
- Updated database and email helper functions to support new notification features, ensuring seamless integration with existing workflows.

vineyy17 committed:
feat: Standardize modal styling by removing box shadows

- Updated ReceiveFundsModal, SendFundsModal, MagicLinkModal, and TransactionPinModal to remove box shadows for a cleaner design.
- Refactored Wallet component to replace user images with Avatar component for improved consistency and user experience.

vineyy17 committed:
feat: Enhance verification code handling with hashing and salting

- Updated session management to store hashed verification codes and salts instead of plain text.
- Introduced HMAC SHA-256 for secure verification code generation and comparison.
- Refactored API routes to support the new verification process, improving security and integrity of user sessions.

vineyy17 committed:
Merge branch 'develop' into feat/wallet-feature

vineyy17 committed:
fix: Ensure contract address type consistency in WithdrawFundsModal

- Updated the contract address parameter in getTokenBalance and getTokenInfo functions to explicitly cast to the expected type, improving type safety and preventing potential runtime errors.

vineyy17 committed:
feat: Enhance transaction details display with meeting information

- Updated TransactionDetailsView to show receiver and host names for debit transactions related to meetings.
- Improved credit transaction display to include sender and guest details when applicable.
- Refactored Wallet component to streamline transaction status and date presentation.
- Modified useWalletTransactions hook to include counterparty names and addresses for better transaction context.
- Adjusted database queries to fetch meeting session data alongside transaction details, enhancing overall user experience.

Onyelaudochukwuka commented on src/pages/api/secure/payment-preferences/index.ts:
```diff
@@ -0,0 +1,80 @@
+    }
+  } else if (req.method === 'PATCH') {
+    const account_address = req.session.account!.address
+    const { owner_account_address, updates } = req.body
```
The account adderess already exists in the backend state

Onyelaudochukwuka commented on src/pages/api/secure/send-change-email-link.ts:
Also add a check to block email updates via /secure/notifications/index

Onyelaudochukwuka commented on src/pages/api/secure/send-change-email-link.ts:
```diff
@@ -0,0 +1,62 @@
+const TOKEN_EXPIRY = '5m'
+
+if (!JWT_SECRET) {
+  throw new Error('JWT_SECRET environment variable is required')
```
A top-level Error would crash the app in the instance that this doesn't exist. Instead, add this check to the api handler

Onyelaudochukwuka commented on src/pages/api/secure/send-change-email-link.ts:
```diff
@@ -0,0 +1,62 @@
+import { sendChangeEmailEmail } from '@/utils/email_helper'
+
+const JWT_SECRET = process.env.JWT_SECRET
+const TOKEN_EXPIRY = '5m'
```
move this to a constant on utils/constants

Onyelaudochukwuka replied:
And be specific with what token. I.e EMAIL_RESET_TOKEN_EXPIRY

Onyelaudochukwuka commented on src/pages/api/secure/send-change-email-link.ts:
```diff
@@ -0,0 +1,62 @@
+    const changeUrl = `${appUrl}/dashboard/change-email?token=${changeToken}&address=${account_address}`
+
+    // Get user's current notification email from the request body
+    const { currentEmail } = req.body
```
The current email is already in the DB and not required on the request; otherwise, a mail can be injected while making the request. The email on the DB is immutable and should be the single source of truth.

Onyelaudochukwuka commented on src/pages/api/secure/send-enable-pin-link.ts:
```diff
@@ -0,0 +1,62 @@
+import { sendEnablePinEmail } from '@/utils/email_helper'
+
+const JWT_SECRET = process.env.JWT_SECRET
+const TOKEN_EXPIRY = '5m'
```
use the constant creted from the feedback above here.

Onyelaudochukwuka commented on src/pages/api/secure/send-reset-pin-link.ts:
```diff
@@ -0,0 +1,63 @@
+const TOKEN_EXPIRY = '5m'
+
+if (!JWT_SECRET) {
+  throw new Error('JWT_SECRET environment variable is required')
```
A top-level Error would crash the app in the instance that this doesn't exist. Instead, add this check to the api handler

Onyelaudochukwuka commented on src/pages/api/secure/send-reset-pin-link.ts:
```diff
@@ -0,0 +1,63 @@
+    const resetUrl = `${appUrl}/dashboard/reset-pin?token=${resetToken}&address=${account_address}`
+
+    // Get user's notification email from the request body
+    const { email } = req.body
```
Again, an email shouldn't be passed from the request. The one on the DB should be the single source of truth for this account.

Onyelaudochukwuka commented on src/pages/api/secure/send-verification-code.ts:
```diff
@@ -0,0 +1,54 @@
+    const hash = hmacSha256(verificationCode, salt)
+
+    req.session.verificationCodeHash = hash
+    req.session.verificationCodeSalt = salt
```
after the user logs out how do we get the salt ?

Onyelaudochukwuka commented on src/pages/api/secure/send-verification-code.ts:
```diff
@@ -0,0 +1,54 @@
+import { withSessionRoute } from '@/ironAuth/withSessionApiRoute'
+import { sendVerificationCodeEmail } from '@/utils/email_helper'
+
+function hmacSha256(value: string, salt: string): string {
```
node:crypto is the worst possible choice we can make for password hashes. Brute forcing pins, which are just 5 digits, would be easy. Replace all pin hashes with argon2 and add a salt secret as an ENV variable, so you save. PIN + SALT. This salt should be a 32-bit hex.

Onyelaudochukwuka commented on src/pages/api/secure/verify-token.ts:
```diff
@@ -0,0 +1,61 @@
+const JWT_SECRET = process.env.JWT_SECRET
+
+if (!JWT_SECRET) {
+  throw new Error('JWT_SECRET environment variable is required')
```
Remove

Onyelaudochukwuka commented on src/pages/api/secure/verify-verification-code.ts:
```diff
@@ -0,0 +1,91 @@
+import { withSessionRoute } from '@/ironAuth/withSessionApiRoute'
+
+function hmacSha256(value: string, salt: string, secret: string): Buffer {
+  return createHmac('sha256', secret).update(`${value}:${salt}`).digest()
```
repetitive code ?

Onyelaudochukwuka commented on src/pages/dashboard/change-email.tsx:
```diff
@@ -0,0 +1,185 @@
+      if (token && address && typeof token === 'string') {
+        try {
+          // Verify the JWT token
+          const decodedToken = await verifyToken(token)
```
Why are we verifying the token first?

Onyelaudochukwuka commented on src/pages/dashboard/change-email.tsx:
```diff
@@ -0,0 +1,185 @@
+        ),
+      }
+
+      return await setNotificationSubscriptions(updatedSubs)
```
Why are we using the same endpoint to update emails ?

Onyelaudochukwuka commented on src/pages/dashboard/enable-pin.tsx:
```diff
@@ -0,0 +1,170 @@
+      }
+
+      return await savePaymentPreferences(address, {
+        pin_hash: pin, // Will be hashed on server
```
If PIN were hashed on the server, why is this called PIN hash? Why can I send the PIN to the server without the token that holds the update state ?

Onyelaudochukwuka commented on src/pages/dashboard/enable-pin.tsx:
```diff
@@ -0,0 +1,170 @@
+      if (token && address && typeof token === 'string') {
+        try {
+          // Verify the JWT token
+          const decodedToken = await verifyToken(token)
```
Let's remove this.

Onyelaudochukwuka commented on src/pages/dashboard/reset-pin.tsx:
```diff
@@ -0,0 +1,169 @@
+      }
+
+      return await savePaymentPreferences(address, {
+        pin_hash: newPin, // Will be hashed on server
```
this shouldn't be called pin has unless the pin is actually hashed.

vineyy17 commented on package.json:
```diff
@@ -62,20 +63,22 @@
     "ethereumjs-util": "^7.1.3",
     "ethers": "5",
     "framer-motion": "^10.12.12",
+    "gaxios": "^7.1.1",
```
Didn't install that. Check the code and see where it's used. I can see you're the one that added it to the sections of code that use it.

vineyy17 commented on src/utils/transaction.helper.ts:
```diff
@@ -6,6 +6,15 @@ import { TransactionCouldBeNotFoundError } from '@/utils/errors'
 
+// Helper function to extract receiver address from ERC20 transfer input data
+function extractReceiverFromERC20Input(input: string): string | null {
+  if (input.startsWith('0xa9059cbb') && input.length >= 138) {
```
Nahh, it's the standard for ERC20, not specific to USDC

vineyy17 committed:
feat: Update payment preferences and email handling for enhanced security and functionality

- Replaced bcrypt with argon2 for hashing PINs, improving security.
- Refactored payment preferences API to streamline creation and updates.
- Introduced new API routes for email change requests and verification, enhancing user experience.
- Updated email notification logic to support new features and ensure consistency.
- Standardized modal styling by removing box shadows across various components for a cleaner design.

vineyy17 committed:
chore: revert changes in discord.helper.ts

vineyy17 committed:
refactor: Simplify hash extraction in Settings component

- Removed unused useColorModeValue import for cleaner code.
- Updated hash extraction logic to use router's asPath for improved readability and reliability.

vineyy17 committed:
feat: Implement verification code handling for enhanced security

- Introduced a new verification system for email and PIN changes, ensuring codes are securely hashed and stored.
- Added API routes for verifying codes and creating verification entries in the database.
- Updated existing routes to utilize the new verification logic, improving overall security and user experience.
- Enhanced error handling for verification processes to provide clearer feedback to users.

vineyy17 committed:
feat: Implement currency conversion and formatting in transaction views

- Added currency conversion functionality to TransactionDetailsView and Wallet components, allowing users to view amounts in their selected currency.
- Introduced utility functions for currency formatting and conversion, enhancing the display of transaction amounts.
- Integrated currency exchange rate fetching using the CurrencyService, improving the accuracy of displayed values.
- Updated relevant components to support dynamic currency selection and display, enhancing user experience.

vineyy17 committed:
feat: Add search functionality to wallet transactions

- Enhanced Wallet component to support search queries for filtering transactions.
- Updated useWalletTransactions hook to accept searchQuery parameter for transaction retrieval.
- Modified API and database functions to handle search queries, improving transaction filtering capabilities.
- Refactored transaction rendering logic to display results based on search criteria, enhancing user experience.

vineyy17 committed:
Merge branch 'develop' into feat/wallet-feature

Onyelaudochukwuka commented on src/pages/api/secure/preferences/payment/reset-pin.ts:
```diff
@@ -0,0 +1,89 @@
+  if (!JWT_SECRET) {
+    return res
+      .status(500)
+      .json({ error: 'JWT_SECRET environment variable is required' })
```
Not useful to users, We should not mention JWT to users. This should be "There was an issue scheduling updating preferences. Please get in touch with us through support@meetwith.xyz"

Onyelaudochukwuka commented on src/pages/api/secure/preferences/payment/index.ts:
```diff
@@ -0,0 +1,88 @@
+    try {
+      // Hash the PIN if it's provided
+      if (updates.pin && typeof updates.pin === 'string') {
+        updates.pin_hash = await createPinHash(updates.pin)
```
Why can a users pin be updated without a token ? Or a confirmation of the old pin on the server ?

vineyy17 replied:
The pin update flow verifies the old PIN on the server before allowing changes. Here's how:

` // Change PIN mutation
  const changePinMutation = useMutation(
    async ({ oldPin, newPin }: { oldPin: string; newPin: string }) => {
      // First verify the old PIN
      const verification = await verifyPin(oldPin)

      if (!verification.valid) {
        throw new Error('The current pin you entered is incorrect')
      }

      // PIN verified, now update with new PIN
      return await updatePaymentPreferences(currentAccount.address, {
        pin: newPin, // Will be hashed on server
      })
    },`
    
    We have a different endpoint for verifying the pin, because the `updatePaymentPreferences` function also handles other preferences asides the transaction pin, went this way to keep things tidy. The `updatePaymentPreferences` will only be called when the pin has been verified on the server.

Onyelaudochukwuka replied:
This is only verifying the change on the client side. The old pin should still be required to verify the change on the server.

Onyelaudochukwuka replied:
This only verifies the change on the client side the pin should still be required to verify the change on the server.

Onyelaudochukwuka replied:
If we don't put such measures in place, someone can skip the verify step, and the pin can still be updated.

Onyelaudochukwuka commented on src/pages/api/secure/preferences/payment/enable-pin.ts:
```diff
@@ -0,0 +1,106 @@
+  if (!JWT_SECRET) {
+    return res
+      .status(500)
+      .json({ error: 'JWT_SECRET environment variable is required' })
```
Replace with the afromentioned error message

Onyelaudochukwuka commented on src/utils/database.ts:
```diff
@@ -5434,6 +5591,262 @@ const recordOffRampTransaction = async (event: OnrampMoneyWebhook) => {
+      .eq('channel', channel)
+      .gte('expires_at', new Date().toISOString())
+      .order('created_at', { ascending: false })
+      .limit(1)
```
what if multiple exists ? say the user generated 2 and used the first one he/she generated?

Onyelaudochukwuka commented on src/utils/database.ts:
```diff
@@ -5434,6 +5591,262 @@ const recordOffRampTransaction = async (event: OnrampMoneyWebhook) => {
+  try {
+    const { error } = await db.supabase
+      .from('verifications')
+      .delete()
```
delete should clean up this and all expired ones.

Onyelaudochukwuka commented on src/pages/api/secure/verify-verification-code.ts:
```diff
@@ -0,0 +1,96 @@
+import { createHmac } from 'crypto'
```
1. Why is this still needed ?

Onyelaudochukwuka replied:
2. Why is this still name the same way ?

Onyelaudochukwuka commented on src/pages/api/secure/accounts/change-email.ts:
```diff
@@ -0,0 +1,96 @@
+      account_address
+    )
+
+    const updatedNotifications = {
```
What if no email channel exists ?

vineyy17 replied:
Added a check for that

Onyelaudochukwuka commented on src/pages/api/secure/notifications/pin/enable.ts:
```diff
@@ -0,0 +1,74 @@
+      expiresAt
+    )
+
+    const enableUrl = `${appUrl}/dashboard/enable-pin?token=${enableToken}&address=${account_address}`
```
why are you passing the account address on the URL ?

Onyelaudochukwuka commented on src/pages/dashboard/change-email.tsx:
```diff
@@ -0,0 +1,110 @@
+
+const ChangeEmailPage = () => {
+  const router = useRouter()
+  const { token, address } = router.query
```
Why are we using the address from a query ? The user needs to be logged in to do this , yes ?

Onyelaudochukwuka commented on src/pages/dashboard/enable-pin.tsx:
```diff
@@ -0,0 +1,111 @@
+
+const EnablePinPage = () => {
+  const router = useRouter()
+  const { token, address } = router.query
```
Why are we using the address from a query ? The user needs to be logged in to do this , yes ?

Onyelaudochukwuka commented on src/pages/dashboard/reset-pin.tsx:
```diff
@@ -0,0 +1,111 @@
+
+const ResetPinPage = () => {
+  const router = useRouter()
+  const { token, address } = router.query
```
Why are we using the address from a query ? The user needs to be logged in to do this , yes ?

Onyelaudochukwuka commented on src/utils/constants.ts:
```diff
@@ -30,14 +37,66 @@ export enum OnboardingSubject {
+  RECEIVE_TOKENS = 'receive-tokens',
+}
+
+export enum CurrencySymbol {
```
why are defining currency symbols when Intl already gices us this ? Why not use the existing parse currency helper ?

Onyelaudochukwuka commented on src/utils/database.ts:
```diff
@@ -4795,10 +4927,26 @@ const createCryptoTransaction = async (
+  let receiverAddress: string | null = null
+
+  try {
+    const feeDetails = await getTransactionFeeThirdweb(
```
This is a verification step which we should not skip

vineyy17 replied:
We're not skipping that. I added it in a trycatch cause sometimes the thirdweb RPC request fails, and at this point the transaction is already completed. In cases where the request fails, the user sees a transaction failed error because the transaction would not be able to be inserted in the db (which isn't a fair reflection of what actually happened). So I added a fallback incase the RPC request fails, so we still have the transaction added to the db when it has indeed been completed.

Onyelaudochukwuka commented on src/utils/transaction.helper.ts:
```diff
@@ -6,6 +6,15 @@ import { TransactionCouldBeNotFoundError } from '@/utils/errors'
 
+// Helper function to extract receiver address from ERC20 transfer input data
+function extractReceiverFromERC20Input(input: string): string | null {
+  if (input.startsWith('0xa9059cbb') && input.length >= 138) {
```
Make '0xa9059cbb' a constant called TRANSFER_FUNCTION_SELECTOR

Onyelaudochukwuka commented on src/components/profile/components/NavMenu.tsx:
```diff
@@ -88,46 +90,37 @@ export const NavMenu: React.FC<{
         icon: FaCalendarAlt,
         mode: EditMode.AVAILABILITY,
       },
+      // {
```
delete the commented lines

Onyelaudochukwuka commented on src/components/profile/AccountDetails.tsx:
```diff
@@ -139,29 +154,88 @@ const AccountDetails: React.FC<{ currentAccount: Account }> = ({
     const subscriptions = await syncSubscriptions()
     currentAccount!.subscriptions = subscriptions
-    login(currentAccount!)
-    setCurrentPlan(isProAccount(currentAccount!) ? Plan.PRO : undefined)
```
Why was this removed ?

Onyelaudochukwuka commented on src/components/profile/AccountDetails.tsx:
```diff
@@ -504,9 +393,15 @@ const AccountDetails: React.FC<{ currentAccount: Account }> = ({
+                  }
                   placeholder={
-                    subscription
+                    currentAccount?.subscriptions?.find(
```
use getActiveProSubscription here that would be more clean

Onyelaudochukwuka commented on src/components/profile/ConnectedAccounts.tsx:
```diff
@@ -153,6 +153,8 @@ const DiscordConnection: React.FC = () => {
           )}&response_type=code&scope=identify%20guilds&state=${Buffer.from(
             JSON.stringify({ origin: OnboardingSubject.DiscordConnectedInPage })
           ).toString('base64')}`}
+          target="_blank"
```
this should open on the same page as the user would get redirected back when they complete signing in to discord.

Onyelaudochukwuka replied:
I.e take out the target : _blank

Onyelaudochukwuka commented on src/ironAuth/withSessionApiRoute.ts:
```diff
@@ -8,6 +8,10 @@ import {
 declare module 'iron-session' {
   interface IronSessionData {
     account?: AccountSession
+    verificationCodeHash?: string
```
remove these

vineyy17 committed:
feat: Enhance error handling and notification management

- Replaced direct error toast displays with a centralized error handling function in NotificationConfig, SendFundsModal, and Wallet components for improved consistency.
- Updated AccountDetails to manage subscription plans more effectively, integrating the current plan state.
- Refactored currency formatting logic in TransactionDetailsView and Wallet components to utilize a new utility function for better maintainability.
- Cleaned up unused code and improved the handling of verification processes in various API routes, ensuring clearer error messages and better user feedback.

vineyy17 committed:
feat: Integrate smart wallet reconnection in SendFundsModal and WithdrawFundsModal

- Added useSmartReconnect hook to both SendFundsModal and WithdrawFundsModal for improved wallet reconnection handling.
- Implemented checks for wallet reconnection status, displaying error messages when reconnection fails, enhancing user experience during fund transactions.
- Updated error handling in API routes to capture exceptions with Sentry for better monitoring and debugging.

vineyy17 committed:
Merge branch 'develop' into feat/wallet-feature

davecowart commented on src/utils/services/discord.helper.ts:
Confirmed and fixed in `73bbd40bf`.

Verified all three legs against source: `generateContractDocument` (`contract.activity.ts:292-317`) inserts the asset via `storageService.uploadFile`, awaits `deleteDocumentApprovalsByDeal`, and only then writes `contractDocumentKey` — non-transactional postCommit, so the window is real and includes a DB round trip. The stop predicate did key on the row alone, and the catch-up effect invalidates `dealKeys.detail(dealId)` only.

Took your first option (gate the stop on the answer), not the second. The `invalidateQueries` route is guarded on `isGeneratingContract && hasGeneratedContractDoc`, and `isGeneratingContract` is `pending_contract && !contractDocumentKey` — the poll comment right above it already records that the deal query usually refetches the key first, which flips that false. In the exact interleaving that strands the cache, the effect can have already stopped firing, so it is not a reliable recovery path.

One thing I checked before taking it, since gating on the answer risks an unbounded 3s poll: the manual-upload path does terminate. Confirming a `contract_upload` calls `dealService.uploadDocument` → `UPLOAD_DOCUMENT`, which writes `contractDocumentKey` (`contract.activity.ts:101`), so the route names the row and the poll stops. Deal status bounds it either way, and the ~20s `generationMayHaveFailed` escape hatch already covers a generation that stalls with a row but no key.

Pulled the predicate out as `contractPollIntervalMs` so it is testable — `DealDetailClient` has no test harness — with a regression test pinning the null-answer-with-a-row case at 3000ms.
